src/plugins/telemetry/public/services/telemetry_service.ts

```diff
@@ -0,0 +1,185 @@
+  private isOptedIn: boolean;
+
+  constructor({
+    http,
+    injectedMetadata,
+    reportOptInStatusChange = true,
+  }: TelemetryServiceConstructor) {
+    const isOptedIn = injectedMetadata.getInjectedVar('telemetryOptedIn') as boolean;
+
+    this.reportOptInStatusChange = reportOptInStatusChange;
+    this.injectedMetadata = injectedMetadata;
+    this.http = http;
+
+    this.isOptedIn = isOptedIn;
+  }
+
+  public getCanChangeOptInStatus() {
+    const allowChangingOptInStatus = this.injectedMetadata.getInjectedVar(
+      'allowChangingOptInStatus'
+    ) as boolean;
+    return allowChangingOptInStatus;
+  }
+
+  public getOptInStatusUrl() {
+    const telemetryOptInStatusUrl = this.injectedMetadata.getInjectedVar(
+      'telemetryOptInStatusUrl'
+    ) as string;
+    return telemetryOptInStatusUrl;
+  }
+
+  public getTelemetryUrl() {
+    const telemetryUrl = this.injectedMetadata.getInjectedVar('telemetryUrl') as string;
+    return telemetryUrl;
+  }
+
+  public getIsOptedIn() {
+    return this.isOptedIn;
+  }
+
+  public async fetchTelemetry({ unencrypted = false } = {}) {
+    const now = moment();
+    return this.http.post(`/api/telemetry/v2/clusters/_stats`, {
+      body: JSON.stringify({
+        unencrypted,
+        timeRange: {
+          min: now.subtract(20, 'minutes').toISOString(),
+          max: now.toISOString(),
+        },
+      }),
+    });
+  }
+
+  public async setOptIn(optedIn: boolean): Promise<boolean> {
+    const canChangeOptInStatus = this.getCanChangeOptInStatus();
+    if (!canChangeOptInStatus) {
+      return false;
+    }
+
+    try {
+      await this.http.post('/api/telemetry/v2/optIn', {
+        body: JSON.stringify({ enabled: optedIn }),
+      });
+      if (this.reportOptInStatusChange) {
+        await this.reportOptInStatus(optedIn);
+      }
+      this.isOptedIn = optedIn;
+    } catch (err) {
+      toastNotifications.addError(err, {
+        title: i18n.translate('telemetry.optInErrorToastTitle', {
+          defaultMessage: 'Error',
+        }),
+        toastMessage: i18n.translate('telemetry.optInErrorToastText', {
+          defaultMessage: 'An error occurred while trying to set the usage statistics preference.',
+        }),
+      });
+
+      return false;
+    }
+
+    return true;
+  }
+
+  private async reportOptInStatus(OptInStatus: boolean) {
+    const telemetryOptInStatusUrl = this.getOptInStatusUrl();
+
+    try {
+      // const optInStatus = await this.http.post('/api/telemetry/v2/clusters/_opt_in_stats', {
+      //   body: JSON.stringify({ enabled, unencrypted: false })
+      // });
+
+      return await fetch(telemetryOptInStatusUrl, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({ enabled: OptInStatus }),
+      });
+    } catch (err) {
+      // Sending the ping is best-effort. Telemetry tries to send the ping once and discards it immediately if sending fails.
```

Comment:
I don't think we wants commented out code in master ;-)

Reply:
Still a `[draft]` so the PR is not ready for a review. Some cases commented code is needed to clarify certain things, like why we are intentionally swallowing errors in this case.